### PR TITLE
fix dependency order in setup_or_migrate.rake

### PR DESCRIPTION
**Pull request by wjordan**
Loading the Rails environment may require database access to complete, so the environment-loading action should be placed within the database_exists? block.
This ensures that `rake db:setup_or_migrate` will continue on to create the database when it doesn't yet exist and the environment-loading action fails.

Cherry-pick+rebased this commit from integration-test work in #6646, as it was necessary to get Chef to fully provision a new full-stack server in the adhoc environment. This should also fix the "rake install fails on initial workstation setup" issue as well.

### CODE REVIEW

File: dashboard/lib/tasks/setup_or_migrate.rake [hunk 1]
```diff
@@ -9,8 +10,10 @@ namespace :db do
     true
   end
 
-  task :setup_or_migrate => [:environment, :load_config] do
-    if database_exists?
+  task :setup_or_migrate do
+    db_exists = database_exists?
+    Rake::Task["db:load_config"].invoke
```
Comment (islemaster) [inline dashboard/lib/tasks/setup_or_migrate.rake:15]: Do I understand correctly that you've moved `db:load_config` out of the task dependencies to preserve task ordering (make sure it runs after `environment`)?

That seems over-cautious here.  Either `load_config` depends on `environment`, in which case `load_config` also belongs inside the `database_exists?` check, or it doesn't, in which case it ought to be okay to run first (and leave it as a task dependency).

Unless you have a different (code style?) reason for doing it this way.

Comment (islemaster) [inline dashboard/lib/tasks/setup_or_migrate.rake:15]: I assume we don't have direct control over the `environment` task here - if we did, I'd say the _most_ correct thing to do would be make _it_ depend on some kind of `ensure_db_exists` task, since you've already identified that it depends on the db existing.

File: dashboard/lib/tasks/seed.rake [hunk 2]
```diff
@@ -9,8 +9,7 @@ namespace :seed do
 
   STANFORD_HINTS_FILE = 'config/stanford-hints-bestPath1.tsv'
   STANFORD_HINTS_IMPORTED = 'config/scripts/.hints_imported'
-  file STANFORD_HINTS_IMPORTED => STANFORD_HINTS_FILE do
-    require_relative '../../config/environment'
+  file STANFORD_HINTS_IMPORTED => [STANFORD_HINTS_FILE, :environment] do
```
Comment (islemaster) [inline dashboard/lib/tasks/seed.rake:12]: Is this related to the change below somehow?

**Comment by islemaster** [comment]
Thanks for tackling this @wjordan! I suspect you know what you're doing, but I have a few questions so that I can understand your approach before I approve.

File: dashboard/lib/tasks/setup_or_migrate.rake (hunk 1 above)
Comment (wjordan) [inline dashboard/lib/tasks/setup_or_migrate.rake:15]: Correct, the change was to preserve task ordering.

I quite simply don't know if `load_config` depends on `environment` or not, or whether it's order-sensitive (it's possible that `load_config` does not strictly depend on `environment`, but behaves differently if `environment` is invoked after it rather than before).

Because I don't know, in this PR I made an effort to match the exact task-dependency-ordering found in Rails's code for these two underlying tasks:
[`db:create`](https://github.com/rails/rails/blob/master/activerecord/lib/active_record/railties/databases.rake#L26)`=> [:load_config]`
[`db:migrate`](https://github.com/rails/rails/blob/master/activerecord/lib/active_record/railties/databases.rake#L57)`=> [:environment, :load_config]`

Here's what (I think) happens in this PR's `setup_or_migrate` task:
- Database doesn't exist:
  - `database_exists?` invokes `environment` and/or `Base.connection`, catches exception, returns `false` (the `:environment` task may or may not have successfully loaded at this point, depending on whether the specific environment's initializers attempted to establish a database connection or not.)
  - invokes `load_config`
  - invokes `db:create`
- Database does exist:
  - `database_exists?` invokes `environment` and/or `Base.connection`, establishes connection, returns `true`
  - invokes `load_config`
  - invokes `db:migrate`

The assumption was that if Rails specifically constructed its task-dependency-ordering like this, it was possibly (probably?) for a reason, and I wanted to introduce as little modified behavior as possible.

File: dashboard/lib/tasks/seed.rake (hunk 2 above)
Comment (wjordan) [inline dashboard/lib/tasks/seed.rake:12]: yes- if I recall running `rake db:setup_or_migrate seed:all` with the changes in this PR threw an exception on this manual 'require', which I suspected was because it was trying to re-load the `environment` after having loaded it once already. In any case, moving this manual require into the task dependency brings this task in line with the rest of the task definitions and fixed the exception.

**Comment by islemaster** [comment]
Thanks for the explanations.  LGTM!
